src/query/expression/src/evaluator.rs

```diff
@@ -951,13 +951,75 @@ impl<'a> Evaluator<'a> {
         unreachable!("expr is not a set returning function: {expr}")
     }
 
+    fn run_array_reduce(&self, column: &Column, expr: &Expr) -> Result<Scalar> {
+        let col_type = column.data_type();
+        if col_type.is_null() || column.len() < 1 {
+            return Ok(Scalar::Null);
+        }
+        let mut arg0 = column.index(0).unwrap().to_owned();
+        for i in 1..column.len() {
+            let arg1 = column.index(i).unwrap().to_owned();
+            let entries = {
+                vec![
+                    BlockEntry::new(col_type.clone(), Value::Scalar(arg0.clone())),
+                    BlockEntry::new(col_type.clone(), Value::Scalar(arg1)),
+                ]
+            };
+            let block = DataBlock::new(entries, 1);
+            let evaluator = Evaluator::new(&block, self.func_ctx, self.fn_registry);
+            arg0 = evaluator.run(expr)?.as_scalar().unwrap().clone();
+            arg0 = self
+                .run_cast(
+                    None,
+                    &arg0.as_ref().infer_data_type(),
+                    &col_type,
+                    Value::Scalar(arg0.clone()),
+                    None,
+                )?
+                .as_scalar()
+                .unwrap()
+                .clone();
+        }
+        Ok(arg0)
+    }
+
     fn run_lambda(
         &self,
         func_name: &str,
         args: Vec<Value<AnyType>>,
         lambda_expr: &RemoteExpr,
     ) -> Result<Value<AnyType>> {
         let expr = lambda_expr.as_expr(self.fn_registry);
+        // array_reduce differs
+        if func_name == "array_reduce" {
+            match &args[0] {
+                Value::Scalar(s) => match s {
+                    Scalar::Array(c) => {
+                        let result = self.run_array_reduce(c, &expr)?;
+                        return Ok(Value::Scalar(result));
+                    }
+                    _ => unreachable!(),
+                },
+                Value::Column(c) => {
+                    let mut builder =
+                        ColumnBuilder::with_capacity(&expr.data_type().wrap_nullable(), c.len());
```

Comment:
As we cast the data type of expr to the argument inner data type, the data type of builder should be `c.data_type()`